src/views/virtualmachines/migrate/utils.ts

```diff
@@ -0,0 +1,174 @@
+          destinationStorageClass,
+          pvc?.spec?.resources?.requests?.storage,
+        );
+      })
+      .map((dataVolume) =>
+        k8sCreate({ data: dataVolume, model: DataVolumeModel, ns: getNamespace(dataVolume) }),
+      ),
+  );
+
+  const { fulfilledRequests, rejectedRequests } = dataVolumesCreationResults.reduce(
+    (acc, result) => {
+      result.status === 'rejected'
+        ? acc.rejectedRequests.push(result)
+        : acc.fulfilledRequests.push(result);
+
+      return acc;
+    },
+    {
+      fulfilledRequests: [] as PromiseFulfilledResult<V1beta1DataVolume>[],
+      rejectedRequests: [] as PromiseRejectedResult[],
+    },
+  );
+
+  if (!isEmpty(rejectedRequests)) {
+    fulfilledRequests.map((result) =>
+      k8sDelete({
+        model: DataVolumeModel,
+        resource: result.value,
+      }),
+    );
+
+    throw new Error(rejectedRequests?.[0].reason);
+  }
+
+  const dataVolumesCreated = fulfilledRequests.map((result) => result.value);
+
+  return originPVCs.reduce((createdDataVolumeByPVCName, pvc, index) => {
+    createdDataVolumeByPVCName.set(getName(pvc), dataVolumesCreated[index]);
+    return createdDataVolumeByPVCName;
+  }, new Map<string, V1beta1DataVolume>());
+};
+
+const createPatchData = (
+  vm: V1VirtualMachine,
+  createdDataVolumeByPVCName: Map<string, V1beta1DataVolume>,
+) =>
+  getVolumes(vm).reduce((patchArray, volume, volumeIndex) => {
+    if (volume?.persistentVolumeClaim?.claimName) {
+      patchArray.push({
+        op: 'replace',
+        path: `/spec/template/spec/volumes/${volumeIndex}/persistentVolumeClaim/claimName`,
+        value: getName(createdDataVolumeByPVCName.get(volume?.persistentVolumeClaim?.claimName)),
+      });
+    }
+
+    if (volume?.dataVolume?.name) {
+      const destinationDataVolumeName = getName(
+        createdDataVolumeByPVCName.get(volume?.dataVolume?.name),
+      );
+
+      const dataVolumeIndex = getDataVolumeTemplates(vm)?.findIndex(
+        (dataVolumeTemplate) => getName(dataVolumeTemplate) === volume?.dataVolume?.name,
+      );
+
+      patchArray.push({
+        op: 'replace',
+        path: `/spec/template/spec/volumes/${volumeIndex}/dataVolume/name`,
+        value: destinationDataVolumeName,
+      });
+
+      patchArray.push({
+        op: 'replace',
+        path: `/spec/dataVolumeTemplates/${dataVolumeIndex}/metadata/name`,
+        value: destinationDataVolumeName,
+      });
+    }
+    return patchArray;
+  }, [] as Patch[]);
+
+export const migrateVM = async (
+  vm: V1VirtualMachine,
+  originPVCs: IoK8sApiCoreV1PersistentVolumeClaim[],
+  destinationStorageClass: string,
+) => {
+  const createdDataVolumeByPVCName = await createBlankDataVolumes(
+    originPVCs,
+    destinationStorageClass,
+  );
+
+  const patchData = createPatchData(vm, createdDataVolumeByPVCName);
+
+  patchData.push({
+    op: 'add',
+    path: '/spec/updateVolumesStrategy',
+    value: 'migration',
+  });
+
+  try {
+    return k8sPatch({
```

Comment:
why not waiting for it and then return it?

Reply:
why? I don't need to do stuff after the patch. I `await` in the VirtualMachineMigration component